torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,145 @@
+import collections
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+from torch.testing._core import _dispatch_dtypes
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           all_types_and_half,
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           floating_types_and_half,
+                           floating_types,
+                           integral_types,
+                           floating_types_and,
+                           floating_and_complex_types_and,
+                           integral_types_and,
+                           all_types_and,
+                           )
+
+COMPLETE_DTYPES_DISPATCH = (
+    all_types,
+    all_types_and_complex,
+    all_types_and_half,
+    floating_types,
+    floating_and_complex_types,
+    floating_types_and_half,
+    integral_types,
+    complex_types,
+)
+
+EXTENSIBLE_DTYPE_DISPATCH = (
+    all_types_and_complex_and,
+    floating_types_and,
+    floating_and_complex_types_and,
+    integral_types_and,
+    all_types_and,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+# Class to tag the dynamically generated types.
+
+
+class _dynamic_dispatch_dtypes(_dispatch_dtypes):
+    pass
+
+
+def get_supported_dtypes(op, sample_inputs_fn, device_type):
+    # Returns the supported dtypes for the given operator and device_type pair.
+    assert device_type in ['cpu', 'cuda']
+    if not TEST_CUDA and device_type == 'cuda':
+        warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA dtypes could be wrong")
+        return _dynamic_dispatch_dtypes(())
+
+    supported_dtypes = set()
+    for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
+        samples = sample_inputs_fn(op, device_type, dtype, False)
```

Comment:
This needs to be a in a try because the sample input functions don't guarantee they can produce valid inputs for every dtype, I think. They only guarantee they generate valid inputs for supported dtypes. So if the sample inputs function fails to produce samples I think this should assume that dtype isn't supported